File: backend/domain/models/invoice.py
from .baseModel import BaseModel
from .columns import (
    StringColumn, TextColumn, JSONBColumn,
    NumericColumn, DateColumn, EnumColumn
)
from typing import List, Dict, BinaryIO
from datetime import date
from enum import Enum
from sqlalchemy import Index

class EnumInvoiceStatus(Enum):
    ALL = "All"
    TBD = "TBD"
    TO_BE_TRAITED = "A Traiter"
    NEED_TO_CHECK = "Avoiriser"
    TO_BE_INVOICED = "A Facturer"
    INVOICED = "Facturer ticket"
    VALIDATED = "Valider"

class Invoice(BaseModel):
    prefix: str = "INV"
    
    name: str = StringColumn(length=255, nullable=False)
    path: str = TextColumn(nullable=False)
    external_id: str = StringColumn(length=255, nullable=True, index=True, unique=True)
    invoice_number: str = StringColumn(length=255, nullable=True)
    invoice_date: date = DateColumn(nullable=True)
    amount_ht: float = NumericColumn(nullable=True)
    amount_ttc: float = NumericColumn(nullable=True)
    amount_tva: float = NumericColumn(nullable=True)
    issuer_name: str = StringColumn(length=500, nullable=True)
    construction_site_address: str = StringColumn(length=500, nullable=True)
    gc_booking: str = StringColumn(length=255, nullable=True)
    status: str = StringColumn(length=255, nullable=False)
    images: list = JSONBColumn(nullable=True)
    extracted_data: dict = JSONBColumn(nullable=True)
    comments: str = TextColumn(nullable=True)
    
    __table_args__ = (
        Index(
            "ix_invoices_status_created_at_id",
            "status", "created_at", "id"
        ),
        Index(
            "ix_invoices_created_at_id",
            "created_at", "id"
        ),
    )
    
    def __post_init__(self):
        if self.status:
            self.status = EnumInvoiceStatus(self.status)
            
    def __repr__(self):
        return f"Invoice(id={self.id}, external_id={self.external_id}, invoice_number={self.invoice_number})"

File: frontend-app/lib/types/invoice.ts
export interface Invoice {
    id: string;
    name: string;
    external_id: string;
    path: string;
    gc_booking: string;
    comments: string | null;
    created_at: string;
    updated_at: string;
    status: EnumInvoiceStatus;
    invoice_date: string;
    issuer_name: string;
    construction_site_address: string;
    invoice_number: string;
    amount_ht: number;
    amount_ttc: number;
    amount_tva: number;
}

export interface InvoiceExtractedData {
    invoice_date: string;
    issuer_name: string;
    construction_site_address: string;
    invoice_number: string;
    amount_ht: number;
    amount_ttc: number;
    amount_tva: number;
    tva: number;
    predicted_data: {
        _tvas: string[];
    };
    openai_response: {
        gc_booking: string;
        invoice_number: string;
        invoice_date: string;
        amount_ht: number;
        amount_ttc: number;
        amount_tva: number;
        issuer_name: string;
        construction_site_address: string;
    };
}

export interface StatsInvoices {
    total: number;
    tbd: number;
    need_to_check: number;
    to_be_treated: number;
    to_be_invoiced: number;
    invoiced: number;
}

export class EnumInvoiceStatus {
    static readonly ALL = 'All';
    static readonly TBD = 'TBD';
    static readonly TO_BE_TRAITED = 'A Traiter';
    static readonly NEED_TO_CHECK = 'Avoiriser';
    static readonly TO_BE_INVOICED = 'A Facturer';
    static readonly INVOICED = 'Facturer ticket';
    static readonly VALIDATED = 'Valider';

    static getStatusLabel(status: EnumInvoiceStatus): string | undefined {
        return Object.entries(this).find(
            ([, v]) => v === status
        )?.[0];
    }

    static fromString(status: string): EnumInvoiceStatus {
        return this.getStatusLabel(status) as EnumInvoiceStatus;
    }
}

export interface StatusInput {
    label?: string;
}